# SEQC-II deep learning for somatic mutation detection project
Here you can find the information about the following SEQC-II paper:
* Sahraeian SME, et al., Achieving Robust Somatic Mutation Detection with Deep Learning Models Derived from Reference Data Sets of a Cancer Sample, Genome Biology, 2022.

## Data
### SEQC-II cancer reference samples data
* Analysis documentation and latest updates about the SEQC-II somatic mutation project can be found at [https://sites.google.com/view/seqc2](https://sites.google.com/view/seqc2).
* All the raw sequencing data (FASTQ files) generated by the working group were deposited into [**NCBI SRA database: SRP162370**](https://trace.ncbi.nlm.nih.gov/Traces/sra/?study=SRP162370)
* Some BAM files can be found [here](http://bit.ly/seqc2_smwg_data) as well.
 * FTP address: ftp://ftp-trace.ncbi.nlm.nih.gov/ReferenceSamples/seqc/Somatic_Mutation_WG/data/
* The [latest high-confidence somatic reference call set](http://bit.ly/somatic_call_set) can be found at the NCBI FTP site.
 * FTP address: ftp://ftp-trace.ncbi.nlm.nih.gov/ReferenceSamples/seqc/Somatic_Mutation_WG/release

### Training Dataset (synthetic spiked samples) for the deep learning paper
* The spiked sample used to train the deep learning models in the paper as well as the trained models are listed at:
ftp://ftp-trace.ncbi.nlm.nih.gov/ReferenceSamples/seqc/Somatic_Mutation_WG/data/DeepLearning_bams/
* The NeuSomatic network models can be found at `models`.

## Description of training data preparation
### WGS-Spike models

Sample | Tumor | Normal | Purities used
--------------------|----------------------------|------------------------|--------------------------
FD2N_FD1N | synthetic_tumor_FD2N | WGS_FD_N_1 | 100%T/100%N, 100%T/95%N
FD3N_FD2N | synthetic_tumor_FD3N | WGS_FD_N_2 | 100%T/100%N, 100%T/95%N
IL2N_IL1N | synthetic_tumor_IL2N | WGS_IL_N_1 | 100%T/100%N, 100%T/95%N
IL1N_IL2N | synthetic_tumor_IL1N | WGS_IL_N_2 | 100%T/100%N, 100%T/95%N
NS2N_NS1N | synthetic_tumor_NS2N | WGS_NS_N_1 | 100%T/100%N, 100%T/95%N
NS1N_NS2N | synthetic_tumor_NS1N | WGS_NS_N_2 | 100%T/100%N, 100%T/95%N
NV2N_NV1N | synthetic_tumor_NV2N | WGS_NV_N_1 | 100%T/100%N, 100%T/95%N
NV3N_NV2N | synthetic_tumor_NV3N | WGS_NV_N_2 | 100%T/100%N, 100%T/95%N
NS_1-5N_ NS_6-9N | synthetic_tumor_NS_1-5N | Merged bam of WGS_NS_N_6, WGS_NS_N_7, WGS_NS_N_8, WGS_NS_N_9 | 100%T/100%N, 100%T/95%N
NS_5-9N_ NS_1-4N | synthetic_tumor_NS_5-9N | Merged bam of WGS_NS_N_1, WGS_NS_N_2, WGS_NS_N_3, WGS_NS_N_4 | 100%T/100%N, 100%T/95%N


* All sequencing data are aligned with BWA-MEM (v0.7.15) followed by Picard MarkDuplicates.
* To generate 95%N mixture, we downsampled pure normal to 95% (using Picard DownsampleSam) and pure tumor to 5% and merged them (using Picard MergeSamFiles)
* The sequencing data for the pure normal WGS samples are deposited at [NCBI SRA database: SRP162370](https://trace.ncbi.nlm.nih.gov/Traces/sra/?study=SRP162370). The links for each sample can be found at at [https://sites.google.com/view/seqc2](https://sites.google.com/view/seqc2).
* The bam files for synthetic tumors along with the ground truths VCF file with spike-in mutations can be found [here](ftp://ftp-trace.ncbi.nlm.nih.gov/ReferenceSamples/seqc/Somatic_Mutation_WG/data/DeepLearning_bams/).
* The training region is chr1-chr22 of GRCh38.

### WGS-GT50 models

Sample | Tumor | Normal | Purities used
--------------------|----------------------------|------------------------|--------------------------
EA1 | WGS_EA_T_1 | WGS_EA_N_1 | 100%T/100%N, 100%T/95%N, 10%T/100%N 
FD1 | WGS_FD_T_1 | WGS_FD_N_1 | 100%T/100%N, 100%T/95%N, 10%T/100%N 
IL1 | WGS_IL_T_1 | WGS_IL_N_1 | 100%T/100%N, 100%T/95%N, 10%T/100%N 
LL1 | WGS_LL_T_1 | WGS_LL_N_1 | 100%T/100%N, 100%T/95%N, 10%T/100%N 
NC1 | WGS_NC_T_1 | WGS_NC_N_1 | 100%T/100%N, 100%T/95%N,10%T/100%N 
NS1 | WGS_NS_T_1 | WGS_NS_N_1 | 100%T/100%N, 100%T/95%N,10%T/100%N 
NV1 | WGS_NV_T_1 | WGS_NV_N_1 | 100%T/100%N, 100%T/95%N, 10%T/100%N 
NS1-9 | WGS_NS_T_combine9 | WGS_NS_N_combine9 | 100%T/100%N, 100%T/95%N, 10%T/100%N 


* All sequencing data are aligned with BWA-MEM (v0.7.15) followed by Picard MarkDuplicates.
* To generate 95%N mixture, we downsampled pure normal to 95% (using Picard DownsampleSam) and pure tumor to 5% and merged them (using Picard MergeSamFiles)
* To generate 10%T mixture, we downsampled pure tumor to 10% (using Picard DownsampleSam) and pure normal to 90% and merged them (using Picard MergeSamFiles)
* The sequencing data for the pure normal WGS samples are deposited at [NCBI SRA database: SRP162370](https://trace.ncbi.nlm.nih.gov/Traces/sra/?study=SRP162370). The links for each sample can be found at at [https://sites.google.com/view/seqc2](https://sites.google.com/view/seqc2).
* The bam files for synthetic tumors along with the ground truths VCF file with spike-in mutations can be found [here](ftp://ftp-trace.ncbi.nlm.nih.gov/ReferenceSamples/seqc/Somatic_Mutation_WG/data/DeepLearning_bams/).
* The ground truth for all these samples are High confidence calls in the SEQC2 reference sample (`data/High.v1.0.vcf.gz`).
* The training region can be found at `data/train_region.bed`.

### WGS-GT50-SpikeWGS10 models
* For this model, we mixed all the training data in the WGS-GT50 models, plus 10% of randomly selected of the training data in WGS-Spike models.

## Evaluation scripts 
* The evaluation scripts used in the deep learning paper can be found at `evaluation_scripts`.
* The test region can be found at `data/test_region_WGS.bed` and `data/test_region_WES.bed`.
